nano/node/confirmation_height_processor.cpp

```diff
@@ -0,0 +1,378 @@
+#include <cassert>
+#include <nano/lib/logger_mt.hpp>
+#include <nano/lib/numbers.hpp>
+#include <nano/lib/utility.hpp>
+#include <nano/node/active_transactions.hpp>
+#include <nano/node/stats.hpp>
+#include <nano/secure/blockstore.hpp>
+#include <nano/secure/common.hpp>
+#include <numeric>
+
+nano::confirmation_height_processor::confirmation_height_processor (nano::pending_confirmation_height & pending_confirmation_height_a, nano::block_store & store_a, nano::stat & stats_a, nano::active_transactions & active_a, nano::block_hash const & epoch_link_a, nano::logger_mt & logger_a) :
+pending_confirmations (pending_confirmation_height_a),
+store (store_a),
+stats (stats_a),
+active (active_a),
+epoch_link (epoch_link_a),
+logger (logger_a),
+thread ([this]() {
+	nano::thread_role::set (nano::thread_role::name::confirmation_height_processing);
+	this->run ();
+})
+{
+}
+
+nano::confirmation_height_processor::~confirmation_height_processor ()
+{
+	stop ();
+}
+
+void nano::confirmation_height_processor::stop ()
+{
+	stopped = true;
+	condition.notify_one ();
+	if (thread.joinable ())
+	{
+		thread.join ();
+	}
+}
+
+void nano::confirmation_height_processor::run ()
+{
+	std::unique_lock<std::mutex> lk (pending_confirmations.mutex);
+	while (!stopped)
+	{
+		if (!pending_confirmations.pending.empty ())
+		{
+			pending_confirmations.current_hash = *pending_confirmations.pending.begin ();
+			pending_confirmations.pending.erase (pending_confirmations.current_hash);
+			// Copy the hash so can be used outside owning the lock
+			auto current_pending_block = pending_confirmations.current_hash;
+			lk.unlock ();
+			add_confirmation_height (current_pending_block);
+			lk.lock ();
+			pending_confirmations.current_hash = 0;
+		}
+		else
+		{
+			condition.wait (lk);
+		}
+	}
+}
+
+void nano::confirmation_height_processor::add (nano::block_hash const & hash_a)
+{
+	{
+		std::lock_guard<std::mutex> lk (pending_confirmations.mutex);
+		pending_confirmations.pending.insert (hash_a);
+	}
+	condition.notify_one ();
+}
+
+// This only check top-level blocks having their confirmation height sets, not anything below
+bool nano::confirmation_height_processor::is_processing_block (nano::block_hash const & hash_a)
+{
+	return pending_confirmations.is_processing_block (hash_a);
+}
+
+/**
+ * For all the blocks below this height which have been implicitly confirmed check if they
+ * are open/receive blocks, and if so follow the source blocks and iteratively repeat to genesis.
+ * To limit write locking and to keep the confirmation height ledger correctly synced, confirmations are
+ * written from the ground upwards in batches.
+ */
+void nano::confirmation_height_processor::add_confirmation_height (nano::block_hash const & hash_a)
+{
+	boost::optional<conf_height_details> receive_details;
+	auto current = hash_a;
+	nano::account_info account_info;
+	std::deque<conf_height_details> pending_writes;
+	release_assert (receive_source_pairs_size == 0);
```

Comment:
Is this worth a release_assert, given it's only used for memory stats?

Reply:
Make it an `assert`